File: muchfloat.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>

#include <immintrin.h>

#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#undef WIN32_LEAN_AND_MEAN

#define DURATION 10000
#define SAMPLE_SIZE 1000

typedef struct _THREAD_RESULT {
	uint64_t operationCount;
	double result;
} THREAD_RESULT;

THREAD_RESULT *results;

DWORD WINAPI DoWork(uint32_t index) {
	uint32_t start = GetTickCount();
	
	__m256d acc = _mm256_set_pd(0.1f, 0.1f, 0.1f, 0.1f);
	__m256d unit = _mm256_set_pd(0.1, 0.1f, 0.1f, 0.1f);
	uint64_t ops = 0;
	
	while (GetTickCount() - start < DURATION) {
		for (uint32_t i = 0; i < SAMPLE_SIZE; i++) {
			acc = _mm256_add_pd(acc, unit);
		}
		ops += SAMPLE_SIZE*4;
	}
	
	double accs[] = {0.1f, 0.1f, 0.1f, 0.1f};
	_mm256_store_pd(accs, acc);
	double sum = accs[0] + accs[1] + accs[2] + accs[3];
	
	results[index].operationCount = ops;
	results[index].result = sum;
	
	return 0;
}

int wmain(int agrc, wchar_t **argv) {
	_putws(L"MuchFloat v0.9");
	
	SYSTEM_INFO sys;
	GetSystemInfo(&sys);
	DWORD procCnt = sys.dwNumberOfProcessors;
	wprintf(L"Number of processors: %u\n", procCnt);
	
	results = malloc(sizeof(THREAD_RESULT)*procCnt);
	
	for(uint32_t i = 0; i < procCnt; i++) {
		CreateThread(NULL, 0, DoWork, i, 0, NULL);
	}
	
	Sleep(DURATION + 1000);
	
	uint64_t sum = 0;
	for(uint32_t i = 0; i < procCnt; i++) {
		wprintf(L"Thread %u: %Iu operations, result was %f\n",
			i, results[i].operationCount, results[i].result);
		sum += results[i].operationCount;
	}
	wprintf(L"Total: %Iu operations\n", sum);
	
	free(results);
}

File: muchfloat.py
from collections import namedtuple
from multiprocessing import cpu_count, Manager, Process
from time import sleep, time

ThreadResult = namedtuple("ThreadResult", ("operation_count", "result"));

duration = 10.0
sample_size = 1000

def do_work(i, results):
	start = time()
	ops = 0
	acc = 0.1
	
	while time() - start < duration:
		for _ in range(sample_size):
			acc += 0.1
		ops += sample_size
	
	results[i] = ThreadResult(ops, acc)

def main():
	global results
	
	print("MuchFloat (Python) v0.9")
	cores = cpu_count()
	print("Number of processors: {}".format(cores))
	
	manager = Manager()
	results = manager.list([None for _ in range(cores)])
	
	jobs = []
	for i in range(cores):
		p = Process(target=do_work, args=(i, results))
		p.start()
		jobs.append(p)
	
	for p in jobs:
		p.join()
	
	sum = 0
	for i, result in enumerate(results):
		print("Thread {}: {} operations, result was {}".format(i+1, result.operation_count, result.result))
		sum += result.operation_count
	print("Total: {} operations".format(sum))

if __name__ == "__main__":
	main()